segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -681,17 +684,24 @@ public void close() {
      * */
     private <R> CompletableFuture<R> executeSerialized(Callable<CompletableFuture<R>> operation, String... segmentNames) {
         Exceptions.checkNotClosed(this.closed.get(), this);
-        return this.taskProcessor.add(Arrays.asList(segmentNames), () -> executeAsync(operation));
+        return this.taskProcessor.add(Arrays.asList(segmentNames), () -> executeExclusive(operation, segmentNames));
     }
 
     /**
-     * Executes the given Callable and returns its result.
+     * Executes the given Callable asynchronously and exclusively.
+     * It returns a CompletableFuture that will be completed with the result.
+     * The operations are not allowed to be concurrent.
      *
-     * @param operation The Callable to execute.
+     * @param operation    The Callable to execute.
      * @param <R>       Return type of the operation.
-     * @return CompletableFuture<R> of the return type of the operation.
-     */
-    private <R> CompletableFuture<R> executeAsync(Callable<CompletableFuture<R>> operation) {
+     * @param segmentNames The names of the Segments involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeExclusive(Callable<CompletableFuture<R>> operation, String... segmentNames) {
+        val shouldRelease = new AtomicBoolean(false);
+        acquire(segmentNames);
+        shouldRelease.set(true);
         return CompletableFuture.completedFuture(null).thenComposeAsync(v -> {
```

Comment:
You can rewrite this as `CompletableFuture.runAsync`. Why do you create a completed future just to have an async continuation?

Reply:
There is no `composeAsync` equivalent to supplyAsync